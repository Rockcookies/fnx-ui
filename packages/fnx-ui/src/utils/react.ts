import { Children, ReactElement } from 'react';
import { isFragment } from 'react-is';

export const toElementArray = (
	children: React.ReactNode,
	option?: { keepEmpty?: boolean },
): ReactElement[] => {
	const keepEmpty = option && option.keepEmpty;
	let ret: React.ReactElement[] = [];

	Children.forEach(children, (child: any) => {
		if ((child === undefined || child === null) && !keepEmpty) {
			return;
		}

		if (Array.isArray(child)) {
			ret = ret.concat(toElementArray(child));
		} else if (isFragment(child) && child.props) {
			ret = ret.concat(toElementArray(child.props.children, option));
		} else {
			ret.push(child);
		}
	});

	return ret;
};
