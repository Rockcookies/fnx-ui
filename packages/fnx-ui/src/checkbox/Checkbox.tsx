import {
	forwardRef,
	useCallback,
	useContext,
	useEffect,
	useImperativeHandle,
	useMemo,
	useRef,
} from 'react';
import useControlledState from '../hooks/use-controlled-state';
import useDefaults from '../hooks/use-defaults';
import useProps from '../hooks/use-props';
import useDefaultsRef from '../hooks/use-defaults-ref';
import { noop } from '../utils/misc';
import { createBEM } from '../utils/namespace';
import useCheckboxRender from './hooks/use-checkbox-render';
import { CheckboxProps, CheckboxRef, CheckboxValue } from './interface';
import { CheckGroupContext } from './context';

const NS = 'fnx-checkbox';
const bem = createBEM(NS);

type CheckboxRequiredProps = Required<
	Pick<CheckboxProps, 'defaultChecked' | 'skipGroup' | 'iconShape'>
>;

const DEFAULT_PROPS: CheckboxRequiredProps = {
	defaultChecked: false,
	skipGroup: false,
	iconShape: 'square',
};

const Checkbox = forwardRef<CheckboxRef, CheckboxProps>((_props, ref) => {
	const [
		{ defaultChecked, skipGroup: _skipGroup, iconShape: _iconShape },
		{
			checked: _checked,
			value: checkValue,
			onChange: _onChange,
			...resetProps
		},
	] = useProps<CheckboxRequiredProps, CheckboxProps>(DEFAULT_PROPS, _props);

	const groupContext = useContext(CheckGroupContext);
	const groupContextRef = useDefaultsRef(groupContext);

	const groupValueList: CheckboxValue[] | undefined = _skipGroup
		? undefined
		: groupContext.value || undefined;
	const skipGroup = groupValueList == null;

	const valueChecked = useMemo(() => {
		if (groupValueList == null) {
			return _checked;
		}

		return checkValue != null
			? groupValueList.includes(checkValue)
			: _checked;
	}, [_checked, checkValue, groupValueList]);

	const { value: checked, onChangeRef } = useControlledState<boolean>({
		value: valueChecked,
		defaultValue: groupValueList != null ? false : defaultChecked,
		onChange: _onChange,
	});

	const rootRef = useRef<HTMLDivElement>(null);
	const prevValueRef = useRef<CheckboxValue | null>(null);
	const disabled = useDefaults<boolean>(
		false,
		resetProps.disabled,
		groupContext.disabled,
	);

	useEffect(() => {
		if (skipGroup || checkValue == null) {
			return;
		}

		const { cancelValue = noop, registerValue = noop } =
			groupContextRef.current;

		if (checkValue !== prevValueRef.current) {
			if (prevValueRef.current != null) {
				cancelValue(prevValueRef.current);
			}

			registerValue(checkValue);
			prevValueRef.current = checkValue;
		}

		return () => {
			cancelValue(checkValue);
		};
	}, [checkValue, groupContextRef, skipGroup]);

	const handleToggle = useCallback(() => {
		if (disabled) {
			return;
		}

		const { toggleValue } = groupContextRef.current;

		if (!skipGroup && toggleValue && checkValue != null) {
			toggleValue(!checked, checkValue);
		} else {
			onChangeRef.current(!checked);
		}
	}, [
		disabled,
		groupContextRef,
		skipGroup,
		checkValue,
		checked,
		onChangeRef,
	]);

	const checkboxRef = useMemo<CheckboxRef>(() => {
		return {
			root: rootRef.current,
			toggle: handleToggle,
		};
	}, [handleToggle]);

	useImperativeHandle<CheckboxRef, CheckboxRef>(ref, () => checkboxRef, [
		checkboxRef,
	]);

	const render = useCheckboxRender(
		{
			role: 'checkbox',
			...resetProps,
			iconShape: _iconShape,
			onClick: (e) => {
				handleToggle();
				resetProps.onClick && resetProps.onClick(e);
			},
		},
		groupContext,
	);

	return render({
		bem,
		checked,
		ref: rootRef,
	});
});

Checkbox.displayName = 'Checkbox';

export default Checkbox;
