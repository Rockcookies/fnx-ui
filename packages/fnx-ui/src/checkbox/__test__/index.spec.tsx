import React from 'react';
import { fireEvent, render, act } from '@testing-library/react';
import TestsDOM from '../../../test/dom';
import TestsReact from '../../../test/react';
import Checkbox, { CheckboxRef, CheckboxGroupRef } from '../';

describe('<Checkbox/>', () => {
	it('disabled prop', async () => {
		const { container } = render(<Checkbox disabled={true}>box1</Checkbox>);
		const icon = TestsDOM.mustQuerySelector(
			container,
			'.fnx-checkbox__icon',
		);

		fireEvent.click(icon);
		expect(
			container.querySelector('.fnx-checkbox__icon--checked'),
		).toBeNull();
	});

	it('labelDisabled prop', async () => {
		const { container, rerender } = render(
			<Checkbox labelDisabled={true}>box1</Checkbox>,
		);
		const icon = TestsDOM.mustQuerySelector(
			container,
			'.fnx-checkbox__icon',
		);
		const label = TestsDOM.mustQuerySelector(
			container,
			'.fnx-checkbox__label',
		);

		fireEvent.click(label);
		expect(
			icon.classList.contains('fnx-checkbox__icon--checked'),
		).toBeFalsy();

		rerender(<Checkbox labelDisabled={false}>box1</Checkbox>);

		fireEvent.click(label);
		expect(
			icon.classList.contains('fnx-checkbox__icon--checked'),
		).toBeTruthy();
	});

	it('iconSize、iconPosition、iconShape、iconCheckedColor', async () => {
		const { container } = render(
			<Checkbox
				defaultChecked={true}
				iconSize={20}
				iconPosition="right"
				iconShape="square"
				iconCheckedColor="red"
			>
				box1
			</Checkbox>,
		);
		const icon = TestsDOM.mustQuerySelector(
			container,
			'.fnx-checkbox__icon',
		);

		expect(icon.getAttribute('style')).toEqual(
			'font-size: 20px; border-color: red; background-color: red;',
		);
		expect(
			icon.classList.contains('fnx-checkbox__icon--square'),
		).toBeTruthy();
	});

	it('icons', () => {
		const { container } = render(
			<Checkbox
				iconShape="plain"
				checkedIcon={<span>checked</span>}
				icon={<span>no-check</span>}
			>
				box1
			</Checkbox>,
		);

		expect(container).toMatchSnapshot();
	});

	it('checkbox onChange', () => {
		const onChange = jest.fn();
		const { container } = render(
			<Checkbox checked={true} defaultChecked={false} onChange={onChange}>
				box1
			</Checkbox>,
		);

		const icon = TestsDOM.mustQuerySelector(
			container,
			'.fnx-checkbox__icon',
		);

		fireEvent.click(icon);
		expect(onChange).toHaveBeenCalledWith(false);
	});

	it('CheckboxRef toggle', () => {
		const onChange = jest.fn();
		const checkboxRef = React.createRef<CheckboxRef>();

		render(
			<Checkbox onChange={onChange} ref={checkboxRef}>
				box1
			</Checkbox>,
		);

		act(() => TestsReact.mustGetRef(checkboxRef).toggle());

		expect(onChange).toHaveBeenCalledWith(true);
	});

	it('skipGroup prop', () => {
		const { container } = render(
			<Checkbox.Group defaultValue={['1']}>
				<Checkbox skipGroup={true} value="1">
					box1
				</Checkbox>
				<Checkbox value="2">box2</Checkbox>
				<Checkbox value="3">box3</Checkbox>
			</Checkbox.Group>,
		);

		const icons = container.querySelectorAll('.fnx-checkbox__icon');
		expect(
			icons[0].classList.contains('fnx-checkbox__icon--checked'),
		).not.toBeTruthy();

		fireEvent.click(icons[1]);
		expect(
			icons[1].classList.contains('fnx-checkbox__icon--checked'),
		).toBeTruthy();
	});

	it('checkboxGroup onChange', () => {
		const onChange = jest.fn();
		const { container } = render(
			<Checkbox.Group onChange={onChange}>
				<Checkbox skipGroup={true} value="1">
					box1
				</Checkbox>
				<Checkbox value="2">box2</Checkbox>
				<Checkbox value="3">box3</Checkbox>
			</Checkbox.Group>,
		);

		const icons = container.querySelectorAll('.fnx-checkbox__icon');
		fireEvent.click(icons[0]);
		expect(onChange).not.toHaveBeenCalled();

		fireEvent.click(icons[1]);
		expect(onChange).toHaveBeenCalledWith(['2']);

		fireEvent.click(icons[2]);
		expect(onChange).toHaveBeenCalledWith(['2', '3']);
	});

	it('maxCheckedCount prop', () => {
		const onChange = jest.fn();
		const { container } = render(
			<Checkbox.Group onChange={onChange} maxCheckedCount={2}>
				<Checkbox value="1">box1</Checkbox>
				<Checkbox value="2">box2</Checkbox>
				<Checkbox value="3">box3</Checkbox>
			</Checkbox.Group>,
		);

		const icons = container.querySelectorAll('.fnx-checkbox__icon');
		fireEvent.click(icons[0]);
		expect(onChange).toHaveBeenCalledWith(['1']);

		fireEvent.click(icons[1]);
		expect(onChange).toHaveBeenCalledWith(['1', '2']);

		fireEvent.click(icons[2]);
		expect(onChange).toHaveBeenCalledTimes(2);
	});

	it('CheckboxGroupRef toggleAll', () => {
		const onChange = jest.fn();
		const groupRef = React.createRef<CheckboxGroupRef>();

		const { container } = render(
			<Checkbox.Group onChange={onChange} ref={groupRef}>
				<Checkbox skipGroup={true} value="1">
					box1
				</Checkbox>
				<Checkbox value="2">box2</Checkbox>
				<Checkbox value="3">box3</Checkbox>
			</Checkbox.Group>,
		);

		act(() => {
			TestsReact.mustGetRef(groupRef).toggleAll({
				checkAll: true,
			});
		});
		expect(onChange).toHaveBeenCalledWith(['2', '3']);

		const icons = container.querySelectorAll('.fnx-checkbox__icon');
		fireEvent.click(icons[1]);

		act(() => TestsReact.mustGetRef(groupRef).toggleAll());
		expect(onChange).toHaveBeenLastCalledWith(['2']);
	});
});
