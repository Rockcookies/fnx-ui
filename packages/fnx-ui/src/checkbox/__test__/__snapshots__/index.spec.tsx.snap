// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox/> icons 1`] = `
<div>
  <div
    aria-checked="false"
    class="fnx-checkbox"
    role="checkbox"
    tabindex="0"
  >
    <span
      class="fnx-checkbox__icon"
    >
      <span>
        no-check
      </span>
    </span>
    <span
      class="fnx-checkbox__label"
    >
      box1
    </span>
  </div>
</div>
`;
